### DISCUSSION TITLE
Typst: Is it possible to rotate output from quarto?

### DISCUSSION THREAD

**Discussion by aghaynes** [quarto-dev/quarto-cli#9705] [Q&A] [answered]
### Description

There are times (e.g. when theres lots of text in a table, or for a figure that works better in landscape, but doesnt quite fit the width of a page) where it would be useful to rotate output from quarto in typst (and probably latex documents).

```` qmd
---
title: "Untitled"
format: typst
keep-typ: true
---

```{r}
#| label: fig-s
#| fig-cap: something
mtcars |> 
  head() |> 
  tinytable::tt()
```
````

It can be done by wrapping the figure block in rotate:

```
#rotate(-90deg, reflow: true, figure([
#[
#let nhead = 1;
...
```
instead of `#figure([`. After making that small change manually and recompiling the PDF:
![image](https://github.com/quarto-dev/quarto-cli/assets/6594028/1cbdca42-9c2c-46e8-a935-0da17ac79473)

This is relatively simple in latex by putting latex blocks before and after to open/close a landscape environment. 

Is it possible to do something for typst?

**Comment by mcanouil** [comment]
That's a Typst question.
You need to look into Typst documentation: https://typst.app/docs

Possibly, you gonna need a Lua filter to modify the cell output.


The answer is the same as previous ones for about the same question and other formats, *i.e.*, use language specific markup:
- https://github.com/quarto-dev/quarto-cli/discussions/8620
- https://github.com/quarto-dev/quarto-cli/discussions/6104
- https://github.com/quarto-dev/quarto-cli/discussions/2791
- https://github.com/quarto-dev/quarto-cli/discussions/680

FYI for changing orientation:

````qmd
---
title: "Quarto Playground"
format: typst
---

{{< lipsum 1 >}}

```{=typst}
#set page(flipped: true)
```

{{< lipsum 1 >}}

```{=typst}
#set page(flipped: false)
```

{{< lipsum 1 >}}
````

**Reply by aghaynes** [reply]
Thanks, but `#set page` flips the entire page, including any header, footer etc. I only want to flip e.g. a table, leaving logos etc in the same place they were. 

The way to do that is with the rotate function... which would involve editing the code output from quarto i suppose (converting `#figure([` to `#rotate(-90deg, reflow: true, figure([` and adding an addition closing `)`)?

**Reply by mcanouil** [reply, marked as answer]
As suggested (I had some time to play a bit):

<img width="724" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/d64b916b-9a8c-4646-9b1e-24d4644266da">


````qmd
---
title: "Quarto Playground"
format: typst
filters:
  - rotate.lua
---

{{< lipsum 1 >}}

```{r}
#| classes: rotate
knitr::kable(head(mtcars[, 1:5]), format = "pipe")
```

{{< lipsum 1 >}}

```{r}
knitr::kable(head(mtcars[, 1:5]), format = "pipe")
```
````

````lua
function Div(div)
  if not div.classes:includes("rotate") then 
    return
  end
  local cell_output = nil
  quarto._quarto.ast.walk(div.content, {
    Div = function(div)
      if div.classes:includes("cell-output-display") then
        cell_output = div
        table.insert(cell_output.content, 1, pandoc.RawInline('typst', "#rotate(-90deg, reflow: true, ["))
        table.insert(cell_output.content, pandoc.RawInline('typst', '])'))
      end
    end
  })

  return cell_output
end
````

**Reply by aghaynes** [reply]
Thats awesome!! Thanks @mcanouil!

**Reply by cscheid** [reply]
... huh we could do some CSS trickery to make this work in HTML as well, and then I'm wondering if this should just be a Quarto feature.

Do we know how to rotate content in DocX?

**Reply by mcanouil** [reply]
I can play a bit more in the next few days.

**Reply by mcanouil** [reply]
Actually `docx` is a nightmare. You need to set `<w:tcPr><w:textDirection w:val="btLr"/></w:tcPr>` in every single cells.
(You have textFlow property for paragraphs, but does not work quite well. Also docx XML is not pptx XML ... https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.textdirection?view=openxml-3.0.1)

The following is the only way from what I gathered for tables:

````openxml
<w:tbl>
  <w:tr>
    <w:tc>
      <w:tcPr>
        <w:textDirection w:val="btLr"/>
      </w:tcPr>
      <w:p>
        <w:r>
          <w:t>Cell 1</w:t>
        </w:r>
      </w:p>
    </w:tc>
    <w:tc>
      <w:tcPr>
        <w:textDirection w:val="btLr"/>
      </w:tcPr>
      <w:p>
        <w:r>
          <w:t>Cell 2</w:t>
        </w:r>
      </w:p>
    </w:tc>
  </w:tr>
</w:tbl>
````

Not that HTML rotation comes with vertical space issue:

<table>
<tr><th>Lua</th><th>HTML</th></tr>
<tr><td>

````lua
local function rotate_typst(div)
  table.insert(div.content, 1, pandoc.RawInline('typst', "#rotate(-90deg, reflow: true, ["))
  table.insert(div.content, pandoc.RawInline('typst', '])'))
  return div
end

local function rotate_html(div)
  table.insert(div.content, 1, pandoc.RawInline('html', "<div style='transform: rotate(-90deg); display: inline-block;'>"))
  table.insert(div.content, pandoc.RawInline('html', '</div>'))
  return div
end

function Div(div)
  if not div.classes:includes("rotate") then 
    return
  end
  local cell_output = nil
  quarto._quarto.ast.walk(div.content, {
    Div = function(div)
      if div.classes:includes("cell-output-display") then
        if FORMAT:match 'html' or FORMAT:match 'revealjs' then
          cell_output = rotate_html(div)
        elseif FORMAT:match 'typst' then
          cell_output = rotate_typst(div)
        else
          cell_output = div
        end
      end
    end
  })

  return cell_output
end

````

</td><td>

<img width="596" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/5a357a3a-142c-4d30-9e30-fcea9db02c4a">

</td></tr>
</table>

**Reply by cscheid** [reply]
Hm, that's unfortunate. Thanks for doing the exploratory work and teaching us about it.

I wonder if it's still worth doing (we have a good number of features that are specific to HTML+PDF, but I'm not in love with adding even more).

**Reply by mcanouil** [reply]
The trick here is the same as for any "enclosing" syntax (my highlight-text extension for instance).
Such as for latex environments or macros, Typst "block" definition, HTML tags, etc.
Maybe we could illustrate LUA filtering for this purpose.

The main logic is the one for the two local functions.

PS: I used the trick I learned from you in a recent filter to build tabset: `quarto._quarto.ast.walk`

**Reply by cderv** [reply]
For Docx, you can also use tricks with objects. If you put the table inside a Text Box, then you can rotate the Text Box and so the table inside it appear rotated. Text Box can have no border and be the size of the table. So it would looks the same I guess. However, I don't know (yet) the Open XML for it.

Though I am not sure this would be that common for Docx to desire a rotated table at the middle of the page. Usually in Word context, what I have seen is wanting the page layout from vertical to horizontal to insert a large table. 

This may be simpler to do that maybe : If someone wants a rotated table, we do a pagebreak, rotate the new page, add the table, page break again, and rotate back the page to keep inserting content. 

Also don't know yet the Open XML for this, but maybe easier than tweaking every single cell. 

I can look deeper if you think one of this worth it

**Reply by cderv** [reply]
> PS: I used the trick I learned from you in a recent filter to build tabset: quarto._quarto.ast.walk 

Why do we need the special internal walk here ? Is usual Pandoc's way `div:walk()` not enough ? 

Asking because if the former needs to always be used, maybe we should put it in a clean user facing part of the API 🤔

**Reply by mcanouil** [reply]
For OpenXML and Docx, you need to use sections, something like below. Then you need to account for Pandoc writer and ensure it does not insert anything in between.

```openxml
<w:sectPr>
  <w:pgSz w:orient="landscape" />
</w:sectPr>
```

But that's not enough, you need also `<w:lastRenderedPageBreak />` in the last paragraph before that.

For instance, a full section define by Word:

````openxml
<w:sectPr w:rsidR="00A5198E" w:rsidSect="00A5198E">
  <w:pgSz w:w="15840" w:h="12240" w:orient="landscape" />
  <w:pgMar w:top="1440" w:right="1440" w:bottom="1440" w:left="1440" w:header="720" w:footer="720"
    w:gutter="0" />
  <w:cols w:space="720" />
  <w:docGrid w:linePitch="326" />
</w:sectPr>
````

<details><summary><i>A "working" Docx example if you like setting page size in a weird unit and want to learn how to control sections and orientations.</i></summary>

````qmd
---
title: "Quarto Playground"
format: docx
---

{{< lipsum 1 >}}

{{< lipsum 1 >}}

```{=openxml}
<w:p>
  <w:pPr>
    <w:sectPr>
      <w:type w:val="nextPage" />
      <w:pgSz w:w="15840" w:h="12240" w:orient="landscape" />
    </w:sectPr>
  </w:pPr>
</w:p>
```

{{< lipsum 1 >}}
````

</details>

As illustrated, that's easy for Typst and LaTeX:

````qmd
---
title: "Quarto Playground"
format: typst
---

{{< lipsum 1 >}}

```{=typst}
#set page(flipped: true)
```

{{< lipsum 1 >}}

```{=typst}
#set page(flipped: false)
```

{{< lipsum 1 >}}

````

````qmd
---
title: "Quarto Playground"
format: pdf
include-in-header:
  text: |
    \usepackage{pdflscape}
---

{{< lipsum 1 >}}

```{=latex}
\begin{landscape}
```

{{< lipsum 1 >}}

```{=latex}
\end{landscape}
```

{{< lipsum 1 >}}
````

**Reply by cderv** [reply]
Ok thanks for the quick example on Landscape. I know **officedown** has this feature, using some Tags sections syntex close to our Shortcode logic I would say : https://ardata-fr.github.io/officeverse/officedown-for-word.html#tags-for-sections

There is one to be replaced by open xml content for Landscape page. We could compare the syntax they use. 

Also **officedown** feature for Quarto is asked by R Users 😉 So it would be interesting to see if some feature can make it!

**Reply by mcanouil** [reply]
Well, you still need to compute page size in order to change orientation.
There are no ways around that, that's how it's implemented/defined.
You are left with no other choices other than do what is required:

- https://github.com/davidgohel/officer/blob/8a482936f9450ba7b2c90539c15d85edad5f3f39/R/ooxml_run_objects.R#L649-L670

**Reply by aghaynes** [reply]
@mcanouil I just had chance to try out the lua filter you wrote. It works, except that tbl-cap and labels are not included. It looks to me as if the caption is stored in the outer div (not sure if thats the correct term) but isn't then applied in the rotated case...
```
::: {#tbl-base .cell .rotate tbl-cap='blah blah blah'}
::: {.cell-output-display}
```
I dont know why it didn't occur to me earlier, but using a raw typst cell also works, including the caption, so maybe this isn't something that needs the be addressed anyway?

````
```{=typst}
#rotate(-90deg, reflow: true, [
```

```{r}
#| tbl-cap: blah blah blah
#| label: tbl-base
head(mtcars) |> tinytable::tt()
```

```{=typst}
])
```
```` 
Thanks a lot!

**Reply by mcanouil** [reply]
> maybe this isn't something that needs the be addressed anyway?

What do you mean?

The filter I wrote (really quickly) can be updated to be more general if it encounters a div with the `.rotate` class. Here it checks if it's "cell-output". You can use the "native" format to understand what is the representation of a document and use keep-md to get the intermediate markdown to help you.
Did you try to modify the filter to fit your needs? Possibly after reading the documentation on LUA filter (https://quarto.org/docs/extensions/lua.html).

**Reply by aghaynes** [reply]
> What do you mean?

As it's simple enough to accomplish via raw latex or typst cells, maybe it doesnt need a filter solution, although that doesnt solve the issue in word documents of course, so maybe it does? I don't know. 

I had a very quick go at modifying your filter and failed, but i'm not at all familiar with lua. With a use case and an example, maybe I'll spend some time to play and figure it out.
